### DISCUSSION TITLE
Using createDerivedMaterial to create a custom picking material for troika text

### DISCUSSION THREAD

**Discussion by bennlich** [protectwise/troika#177] [Q&A]
Hiya! I'm trying to use troika three utils' `createDerivedMaterial` to create a custom picking material for my troika text mesh. My first attempt was to render a solid color on the troika text surface by using a trivial custom fragment shader:

```js
const troikaTextMesh = new troikaThreeText.Text()
troikaTextMesh.text = 'some text'

// Create a custom picking material that the GPU picker can use
let troikaMaterial = troikaTextMesh.material
let pickingMaterial = troikaThreeUtils.createDerivedMaterial(troikaMaterial, {
  customRewriter: ({ vertexShader, fragmentShader }) => {
    fragmentShader = `
      uniform vec4 objectId;
      void main() {
        gl_FragColor = vec4(1.0, 0.0, 0.0, 1.0);
      }
    `
    return { vertexShader, fragmentShader }
  }
})

let dummyMesh = new THREE.Mesh(troikaTextMesh.geometry, pickingMaterial)
```

I was expecting my dummyMesh to appear as a solid red plane where my text was, but I don't see anything :-/

I wonder if I'm missing something obvious? I'd love any tips for how to debug. Thank you!

**Comment by lojjic** [comment]
I haven't tried it out yet, but it seems like that approach should probably work? The only thing I notice is `troikaTextMesh.pickingMaterial` isn't defined in your example code. There may be some other details around when/how the uniforms needed by the vertex shader get updated.

**Reply by bennlich** [reply]
> The only thing I notice is troikaTextMesh.pickingMaterial isn't defined in your example code.

Thanks-- was a typo just in the example--I edited it away.

> There may be some other details around when/how the uniforms needed by the vertex shader get updated.

Hmm ya. This is what I was wondering. I guess if I dig around in the troika text source I might find something.

**Reply by bennlich** [reply]
I think I found the missing piece for updating the uniforms: `_prepareForRender(material)`, normally called in Text.onBeforeRender.

```js
const myText = new troikaThreeText.Text()
myText.text = "Once upon a time, a big purple dragon was cooking some homemade chicken soup."
myText.fontSize = 5
myText.maxWidth = 100
myText.position.y = 10
myText.color = 0x9966FF

myText.sync(() => {
  // _prepareForRender is important. It updates all the uniforms.
  myText._prepareForRender(myText.material)
  // this mesh gets rendered
  world.scene.add(new THREE.Mesh(myText.geometry, myText.material))
})
```

**Reply by bennlich** [reply]
Heh. Screenshot of my derived material below. I'm surprised that each glyph has its own rectangle. I was expecting to see one giant red rectangle that bounds the whole paragraph. I guess that's not how troika three text vertex shader works.

```js
const myText = new troikaThreeText.Text()
myText.text = "Once upon a time, a big purple dragon was cooking some homemade chicken soup.\n\n\"I'll be damned!\" said the dragon."
myText.font = "/fonts/Chango-Regular.ttf"
myText.fontSize = 5
myText.maxWidth = 100
myText.position.y = 10
myText.color = 0x9966FF

// Update the rendering:
myText.sync(() => {
  // _prepareForRender is important. It updates all the uniforms.
  myText._prepareForRender(myText.material)

  let derivedMaterial = troikaThreeUtils.createDerivedMaterial(myText.material, {
    customRewriter: ({ vertexShader, fragmentShader }) => {
      fragmentShader = `
        uniform vec4 objectId;
        void main() {
          gl_FragColor = vec4(1.0, 0.0, 0.0, 1.0);
        }
      `
      return { vertexShader, fragmentShader }
    }
  })

  let textCopy = new THREE.Mesh(myText.geometry, derivedMaterial)
  textCopy.position.y = 20
  world.scene.add(textCopy)
})
```

![Screenshot from 2022-01-13 18-27-20](https://user-images.githubusercontent.com/704368/149441161-033232be-7b3a-4a17-9006-0e0e875105c3.png)

**Reply by lojjic** [reply]
Very nice! I was actually going to ask if you might rather have the single rectangle. If so, this can probably be quite a bit simpler -- just create a PlaneGeometry+MeshBasicMaterial and sync its position/scale to the Text instance's bounds.

**Reply by bennlich** [reply]
I was thinking about that, and then I was wondering if it would be trickier in the curved case. Do you think that would also be pretty straightforward?

**Reply by lojjic** [reply]
Oh good point, I forgot about curved.

The [vertex shader snippet for that](https://github.com/protectwise/troika/blob/7bf9c5caa222009370eec2b6c725fdc4d5a2f975/packages/troika-three-text/src/TextDerivedMaterial.js#L47-L52) is pretty small, wouldn't be too bad to copy?

**Reply by bennlich** [reply]
Oh, nice! That would work. Okay I'll think about how to do this. Thank you!

I'm using this excellently simple gpu picking lib, and opened an issue to discuss: https://github.com/bzztbomb/three_js_gpu_picking/issues/10

**Reply by lojjic** [reply]
Ha that's cool! @bzztbomb and I are work colleagues.

**Reply by bennlich** [reply]
Hah!!!! No way!!

**Reply by bzztbomb** [reply]
@lojjic Heh, funny to bump into you over here! ;)
